_posts/2021-07-07-#22363.md

```diff
@@ -0,0 +1,79 @@
+pr: 22363
+authors: [theStack]
+components: ["tests", "refactor"]
+host: glozow
+status: upcoming
+commit:
+---
+
+## Notes
+
+- Bitcoin transactions encode spending conditions through a _scriptPubKey_ ("locking script") in
+  outputs and a witness and _scriptSig_ ("unlocking scripts") in the inputs. You can read more about
+Bitcoin Script [here](https://en.bitcoin.it/wiki/Script). In the functional test framework, scripts
+are represented using the
+[CScript](https://github.com/bitcoin/bitcoin/blob/2749613020ed96a4e9204f8b98c44524b0a62d0d/test/functional/test_framework/script.py#L423)
+class, and can be initialized using an array of opcodes and byte-encoded data.
+
+- PR [#22363](https://github.com/bitcoin/bitcoin/pull/22363) replaces many manually-constructed
+  default scripts in the functional tests with helper functions provided in `wallet_util.py`. It
+also corrects an error in the helper function,
+[get_multisig](https://github.com/bitcoin/bitcoin/pull/22363/files#diff-cc5214ad594f50da8e14449daae7f8cff3e8d385c623b1ca13efeb3e5ca0446bR86),
+in which the P2SH-wrapped P2WSH script hadn't hashed the witness script before putting it into the
+scriptSig. We'll use this opportunity to review script and output types.
+
+- To test your understanding of scripts and output types, you can try to fill out this table (Hint:
+  a few cells have been pre-filled, and some cells should remain blank):
+
+<img src="../assets/img/script_output_types_blank.png" width="800">
+
+We will fill out this table throughout the review club, and solutions will be posted after the
+meeting.
+
+## Questions
+
+1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
+NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?
+
+2. What do `key_to_p2pkh_script`, `script_to_p2sh_script`, `key_to_p2wpkh_script` and
+    `script_to_p2wsh_script` in wallet\_util.py do? In what cases would we want to use or not use them?
+
+3. Review of Terminology: Let's define script code, witness script, redeem script, scriptPubKey,
+    scriptSig, witness, and witness program (some of these terms are synonymous).
+
+4. What does the operation `OP_HASH160` do? (Hint: what does the [script
+interpreter](https://github.com/bitcoin/bitcoin/blob/2749613020ed96a4e9204f8b98c44524b0a62d0d/src/script/interpreter.cpp#L431)
+    do when it sees this opcode? What are the differences between the
+[hashers](https://github.com/bitcoin/bitcoin/blob/2749613020ed96a4e9204f8b98c44524b0a62d0d/src/hash.h)?)
+
+5. Review of P2PKH: to send coins to someone by public key hash (pre-segwit), what is included in
+    the scriptPubKey of the output? What needs to be provided in the input when the coin is spent?
+
+6. Review of P2SH: to send coins to someone with spending conditions encoded in a script, what is
+    included in the scriptPubKey of the output? What needs to be provided in the input when the coin is
+    spent? Why do we Pay-To-Script-Hash instead of Pay-To-Script?
```

Comment:
```suggestion
    spent? Why do we use Pay-To-Script-Hash instead of Pay-To-Script?
```

(I think we should avoid the use of P2SH, etc as verbs)

Reply:
done!